ImageGalery/Utilities.swift

```diff
@@ -0,0 +1,216 @@
+//
+//  Utilities.swift
+//
+//  Created by CS193p Instructor.
+//  Copyright © 2017 Stanford University. All rights reserved.
+//
+
+import UIKit
+
+class ImageFetcher
+{
+    // Public API
+    
+    // To use, create with the closure you want called when the image is ready.
+    // Example: let fetcher = ImageFetcher() { // code to execute when fetch is done }
+    // Your closure is invoked OFF THE MAIN THREAD.
+    // Then call fetch(url:) with the url you want to fetch.
+    // And set a backup image in case the fetch fails.
+    //
+    // The handler will be called immediately if the fetch succeeds.
+    // If the fetch fails, the handler will be called if and when the backup image is set.
+    // The backup can be set at any time (i.e. before, during or after the fetch).
+    // If the fetch fails and a backup image is never set, the handler will never be called.
+    // Thus it would sort of be a strange use of this class to not set a backup image
+    //   (because you'd never find out when the fetch failed).
+    // Note that you must keep a strong pointer to this object until the fetch finishes
+    //   otherwise the result of the fetch will be discarded and the handler never called.
+    // In other words, keeping a strong pointer to your instance says "I'm still interested in its result."
+    
+    var backup: UIImage? { didSet { callHandlerIfNeeded() } }
+    
+    func fetch(_ url: URL) {
+        DispatchQueue.global(qos: .userInitiated).async { [weak self] in
+            if let data = try? Data(contentsOf: url.imageURL) {
```

Comment:
You should avoid getting into the pyramid of doom (
https://www.google.com/url?sa=i&rct=j&q=&esrc=s&source=images&cd=&cad=rja&uact=8&ved=2ahUKEwj_lJ-D07fcAhWFjKQKHf1OCkkQjRx6BAgBEAU&url=https%3A%2F%2Fcollinmakersquare.wordpress.com%2F2016%2F02%2F17%2Fwk7-day2-promises-and-error-handling%2F&psig=AOvVaw2vmn0yDmfu7DjMxzgZgZbC&ust=1532518646698556)

use a guard let instead, this would be more elegant:
guard let \`self\` = self, 
let data = try? Data(contentsOf: url.imageURL), 
let image = UIImage(data: data) else { 
self?.fetchFailed = true
return
 }
self.handler(url, image)